Samples/NetworkMediaStreamer/NetworkMediaStreamerBase/src/NwMediaStreamSinkBase.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for more information
+
+#include <pch.h>
+
+using namespace winrt;
+
+NwMediaStreamSinkBase::NwMediaStreamSinkBase(IMFMediaType* pMediaType, IMFMediaSink* pParent, DWORD dwStreamID)
+    : m_pVideoHeader(nullptr)
+    , m_VideoHeaderSize(0)
+    , m_bIsShutdown(false)
+    , m_pParentSink(pParent) // to avoid circular reference never Add-Ref to pParent unless giving ownership to another class
+    , m_dwStreamID(dwStreamID)
+{
+    winrt::check_hresult(MFCreateEventQueue(m_spEventQueue.put()));
+    winrt::check_hresult(MFCreateSimpleTypeHandler(m_spMTHandler.put()));
+    winrt::check_hresult(m_spMTHandler->SetCurrentMediaType(pMediaType));
+}
+
+NwMediaStreamSinkBase ::~NwMediaStreamSinkBase()
+{
+    if (m_pVideoHeader != nullptr)
+    {
+        m_VideoHeaderSize = 0;
+        CoTaskMemFree(m_pVideoHeader);
+    }
+}
+
+
+// IMFClockStateSink methods.
+
+STDMETHODIMP NwMediaStreamSinkBase::Start(MFTIME hnsSystemTime, LONGLONG llClockStartOffset)
+{
+    RETURNIFSHUTDOWN;
+    auto hr = QueueEvent(MEStreamSinkStarted, GUID_NULL, S_OK, nullptr);
+    if (m_spMTHandler)
+    {
+        winrt::com_ptr<IMFMediaType> spMT;
+        hr = m_spMTHandler->GetCurrentMediaType(spMT.put());
+        (void)spMT->GetAllocatedBlob(MF_MT_MPEG_SEQUENCE_HEADER, &m_pVideoHeader, &m_VideoHeaderSize);
+    }
+    if (SUCCEEDED(hr))
+    {
+        hr = QueueEvent(MEStreamSinkRequestSample, GUID_NULL, S_OK, nullptr);
+    }
+    return hr;
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::Stop(MFTIME hnsSystemTime)
+{
+    RETURNIFSHUTDOWN;
+    return QueueEvent(MEStreamSinkStopped, GUID_NULL, S_OK, nullptr);
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::Pause(MFTIME hnsSystemTime)
+{
+    RETURNIFSHUTDOWN;
+    return QueueEvent(MEStreamSinkPaused, GUID_NULL, S_OK, nullptr);;
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::Shutdown()
+{
+    RETURNIFSHUTDOWN;
+    m_bIsShutdown = true;
+    return S_OK;
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::BeginGetEvent(IMFAsyncCallback* pCallback, IUnknown* pState)
+{
+    RETURNIFSHUTDOWN;
+    return m_spEventQueue->BeginGetEvent(pCallback, pState);
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::EndGetEvent(IMFAsyncResult* pResult, IMFMediaEvent** ppEvent)
+{
+    RETURNIFSHUTDOWN;
+    return m_spEventQueue->EndGetEvent(pResult, ppEvent);
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::GetEvent(DWORD dwFlags, IMFMediaEvent** ppEvent)
+{
+    RETURNIFSHUTDOWN;
+    return m_spEventQueue->GetEvent(dwFlags, ppEvent);
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::QueueEvent(MediaEventType met, REFGUID extendedType, HRESULT hrStatus, const PROPVARIANT* pvValue)
+{
+    RETURNIFSHUTDOWN;
+    return m_spEventQueue->QueueEventParamVar(met, extendedType, hrStatus, pvValue);
+}
+
+STDMETHODIMP NwMediaStreamSinkBase::GetMediaSink(IMFMediaSink** ppMediaSink)
+{
+    RETURNIFSHUTDOWN;
+
+    if (!ppMediaSink)

```

Comment:
can use RETURN_HR_IF_NULL

Reply:
I cannot use wil, as it needs Nuget. This code will eventually make it to cmake based build in ROS. and importing nugets in cmake is a pain. 

---
In reply to: [494098098](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r494098098) [](ancestors = 494098098)

Reply:
i can define a macro locally.

---
In reply to: [494404354](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r494404354) [](ancestors = 494404354,494098098)